File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> map;
        map[')'] = '(';
        map['}'] = '{';
        map[']'] = '[';

        stack<char> stack;

        for (auto& c : s) {
            if (map[c]) {
                if (!stack.empty() && stack.top() == map[c]) {
                    stack.pop();
                }
                else {
                    return false;
                }
            }

            else {
                stack.push(c);
            }
        }

        if (stack.empty()) {
            return true;
        }

        return false;
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        mp = {
            ')': '(',
            '}': '{',
            ']': '['
        }


        for c in s:
            if c == '(' or c == '{' or c == '[':
                 stack.append(c)

            elif stack and mp[c] == stack[-1]:
                stack.pop()
            
            else:
                return False
        
        return not stack

           
        